File: blockchain/deploy.py
import os
import json
from web3 import Web3
from solcx import compile_standard, install_solc
from dotenv import load_dotenv

# Kết nối Ganache
w3 = Web3(Web3.HTTPProvider("http://127.0.0.1:7545"))
account = w3.eth.accounts[0]
private_key = os.getenv("PRIVATE_KEY") or input("Nhập PRIVATE_KEY của bạn (từ Ganache): ")

# Load và compile .sol
install_solc("0.8.0")
with open(r"D:\foodtrace\sapxong\blockchain\contracts\FoodTraceability.sol", "r") as file:
# with open("contracts/FoodTraceability.sol", "r") as file: #uncommand neu may khong nhan duoc duong dan
    contract_source = file.read()

compiled_sol = compile_standard({
    "language": "Solidity",
    "sources": {
        "FoodTraceability.sol": {
            "content": contract_source
        }
    },
    "settings": {
        "outputSelection": {
            "*": {
                "*": ["abi", "metadata", "evm.bytecode"]
            }
        }
    }
}, solc_version="0.8.0")

# Lấy ABI và bytecode
contract_interface = compiled_sol['contracts']['FoodTraceability.sol']['FoodTraceability']
contract_abi = contract_interface['abi']
bytecode = contract_interface['evm']['bytecode']['object']

# Deploy contract
contract = w3.eth.contract(abi=contract_abi, bytecode=bytecode)
nonce = w3.eth.get_transaction_count(account)
tx = contract.constructor().build_transaction({
    'from': account,
    'nonce': nonce,
    'gas': 3000000,
    'gasPrice': w3.to_wei('20', 'gwei')
})

signed_tx = w3.eth.account.sign_transaction(tx, private_key=private_key)
tx_hash = w3.eth.send_raw_transaction(signed_tx.raw_transaction)
tx_receipt = w3.eth.wait_for_transaction_receipt(tx_hash)

# Save ABI + contract address
with open("FoodTraceability.json", "w") as f:
    json.dump({
        "abi": contract_abi,
        "address": tx_receipt.contractAddress
    }, f, indent=2)

# Save .env
with open(".env", "w") as f:
    f.write(f"PRIVATE_KEY={private_key}\n")
    f.write(f"CONTRACT_ADDRESS={tx_receipt.contractAddress}\n")

print(f"Deployed! Contract Address: {tx_receipt.contractAddress}")

File: backend/scripts/deploy.js
const Web3 = require('web3');
const fs = require('fs');
const path = require('path');
const solc = require('solc');

// Kết nối với Ganache
const web3 = new Web3('http://localhost:8545');

// Đọc contract source
const contractPath = path.resolve(__dirname, '../blockchain/contracts/FoodTraceability.sol');
const contractSource = fs.readFileSync(contractPath, 'utf8');

// Cấu hình biên dịch
const input = {
  language: 'Solidity',
  sources: {
    'FoodTraceability.sol': {
      content: contractSource
    }
  },
  settings: {
    outputSelection: {
      '*': {
        '*': ['*']
      }
    }
  }
};

async function deployContract() {
  try {
    // Biên dịch contract
    const output = JSON.parse(solc.compile(JSON.stringify(input)));
    const contract = output.contracts['FoodTraceability.sol']['FoodTraceability'];
    const abi = contract.abi;
    const bytecode = contract.evm.bytecode.object;

    // Lấy danh sách tài khoản
    const accounts = await web3.eth.getAccounts();
    const deployer = accounts[0];

    console.log('Deploying contract with account:', deployer);

    // Triển khai hợp đồng
    const contractInstance = new web3.eth.Contract(abi);
    const deployTx = contractInstance.deploy({
      data: '0x' + bytecode,
      arguments: []
    });

    const gas = await deployTx.estimateGas();
    const deployedContract = await deployTx.send({
      from: deployer,
      gas: gas
    });

    console.log('Contract deployed at address:', deployedContract.options.address);

    // Lưu địa chỉ hợp đồng vào file .env
    const envPath = path.resolve(__dirname, '../.env');
    let envContent = '';
    
    if (fs.existsSync(envPath)) {
      envContent = fs.readFileSync(envPath, 'utf8');
    }

    // Cập nhật hoặc thêm CONTRACT_ADDRESS
    if (envContent.includes('CONTRACT_ADDRESS=')) {
      envContent = envContent.replace(
        /CONTRACT_ADDRESS=.*/,
        `CONTRACT_ADDRESS=${deployedContract.options.address}`
      );
    } else {
      envContent += `\nCONTRACT_ADDRESS=${deployedContract.options.address}`;
    }

    fs.writeFileSync(envPath, envContent);
    console.log('Contract address saved to .env file');

    // Lưu ABI vào file
    const abiPath = path.resolve(__dirname, '../src/contracts/FoodTraceability.json');
    fs.writeFileSync(abiPath, JSON.stringify(abi, null, 2));
    console.log('Contract ABI saved to file');

  } catch (error) {
    console.error('Error deploying contract:', error);
  }
}

deployContract(); 